File: apps/node/src/chapter02/random.ts
/**
 * シード付きの疑似乱数生成器（mulberry32）。0 以上 1 未満の値を返す関数を作る。
 * 暗号の用途には使えない。
 */
export function createRandom(seed: number): () => number {
  let state = seed >>> 0;
  return () => {
    state = (state + 0x6d2b79f5) >>> 0;
    let z = state;
    z = Math.imul(z ^ (z >>> 15), z | 1);
    z ^= z + Math.imul(z ^ (z >>> 7), z | 61);
    return ((z ^ (z >>> 14)) >>> 0) / 2 ** 32;
  };
}

/** Fisher–Yates のシャッフル。元の配列は変更せず、並べ替えた新しい配列を返す。 */
export function shuffle<T>(items: readonly T[], random: () => number): T[] {
  const result = [...items];
  for (let i = result.length - 1; i > 0; i--) {
    const j = Math.floor(random() * (i + 1));
    [result[i], result[j]] = [result[j] as T, result[i] as T];
  }
  return result;
}

File: apps/dotnet/src/MachineLearning/Chapter02/Random.fs
module MachineLearning.Chapter02.Random

open System

/// シードを使って Fisher–Yates のシャッフルで並べ替えたリストを返す。
/// 同じシードなら同じ順になる。元のリストは変更されない（F# のリストは不変）。
let shuffle (seed: int) (items: 'T list) : 'T list =
    let random = Random seed
    let array = List.toArray items

    for i in array.Length - 1 .. -1 .. 1 do
        let j = random.Next(i + 1)
        let tmp = array[i]
        array[i] <- array[j]
        array[j] <- tmp

    List.ofArray array
